File: web/src/lib/manifest.ts
// Mirrors the baked manifest.json (camelCase, lowercase enums). The client reads the descriptor and
// obeys it — one code path renders every mark and viewport because only the config differs.

export type Viewport = 'geo' | 'orthographic';
export type Mark = 'point' | 'line' | 'arc' | 'rect' | 'polygon' | 'heat' | 'text';

export interface Bbox {
  minX: number;
  minY: number;
  maxX: number;
  maxY: number;
}

export interface ChannelSpec {
  name: string;
  column: string;
  role: 'measure' | 'dimension' | 'temporal' | 'identity';
  type: string;
}

export interface GeometrySpec {
  kind: string;
}

export interface SourceSpec {
  adapter: string;
  query: string;
  geometry: GeometrySpec;
  channels: ChannelSpec[];
}

/** A data channel bound to a visual channel, optionally with a named scheme (e.g. color ramp). */
export interface ChannelRef {
  channel: string;
  scheme?: string;
}

export type ScaleType =
  | 'linear'
  | 'log'
  | 'sqrt'
  | 'diverging'
  | 'quantize'
  | 'quantile'
  | 'threshold'
  | 'ordinal'
  | 'categorical';

/** How a data channel maps to color. A superset scale spec (à la Vega-Lite): pick the `type`, a named
 *  `scheme` or explicit `range`, and the modifiers that type uses. Everything but `channel` is optional
 *  — the client infers a sensible scale from the channel's datatype when omitted. */
export interface ColorSpec {
  channel: string;
  type?: ScaleType;
  scheme?: string;
  range?: string[]; // explicit hex list — overrides scheme
  domain?: (number | string)[]; // numeric [min,max] or explicit category order
  reverse?: boolean;
  midpoint?: number; // diverging
  bins?: number; // quantize / quantile
  thresholds?: number[]; // threshold
  palette?: Record<string, string>; // categorical: explicit value → hex
  unknown?: string; // color for unmapped / null values
}

export interface EncodingSpec {
  color?: ColorSpec;
  size?: ChannelRef;
}

/** Click-to-inspect config. When absent, marks aren't pickable and clicks do nothing. */
export interface InspectSpec {
  channels: string[];
  title?: string;
}

/** A computed per-mark value (VIEW_CONFIG §4). Its presence opts the view into the group regime. */
export interface MeasureSpec {
  name: string;
  expr: string;
}

export interface ViewConfig {
  id: string;
  title?: string;
  viewport: Viewport;
  mark: Mark;
  reduction?: string;
  source: SourceSpec;
  measures?: MeasureSpec[];
  encoding?: EncodingSpec;
  inspect?: InspectSpec;
}

/** Group regime: the derived split of a view's channels. perMark channels ride the render tiles;
 *  perFact channels live only in the fact companions and drive the fold. */
export interface FactChannels {
  perMark: string[];
  perFact: string[];
}

export interface TileMeta {
  z: number;
  x: number;
  y: number;
  count: number;
  isLeaf: boolean;
}

/** One channel's data domain, scanned from the FULL extract at bake time (unlike the root tile, which
 *  is a sample). Numeric channels carry min/max + a quantile grid; others their distinct values —
 *  `valuesTruncated` marks a capped list, which the client treats as absent. */
export interface ChannelDomain {
  values?: string[];
  valuesTruncated?: boolean;
  min?: number;
  max?: number;
  quantiles?: number[];
}

export interface Manifest {
  version: string;
  /** Tile binary format. 2 = zero-copy (views over one buffer); absent/0/1 = the copy-based decode. */
  tileFormat?: number;
  view: ViewConfig;
  /** The reduction the bake planner chose (e.g. 'aggregate', 'quadtreeLod'). Drives the fidelity label. */
  reduction: string;
  regime: string;
  root: Bbox;
  minZoom: number;
  maxZoom: number;
  tilePointBudget: number;
  totalPoints: number;
  tiles: TileMeta[];
  /** Absent on manifests from older bakes — consumers fall back to scanning the root tile. */
  channelDomains?: Record<string, ChannelDomain>;
  /** Group regime only: the perMark/perFact split (routes a filter to a GPU predicate or the fold). */
  factChannels?: FactChannels;
  /** True when the bake wrote a `z/x/y.facts.arrow` companion beside every tile. */
  companionTiles?: boolean;
  /** The companion grain columns (perFact dict + temporal), the dimensions the fact partials key by. */
  grainChannels?: string[];
  /** Companion packaging (companion-scale R2, extended by R1/R5): companion tiles live as independently
   *  compressed blocks in one per-version archive, range-read per tile. Absent selects the per-file layout
   *  (older bakes). A slab bake packs both leaf and internal levels and fills `planeEntries`. */
  companionPack?: CompanionPack;
  /** Group-regime slab companion metadata (companion-scale R1). Present ⇒ the pack holds slab planes and
   *  the client takes the slab decode/fold; absent ⇒ the row-form companion path (older bakes). */
  companionSlab?: CompanionSlab;
  /** Group-regime only (companion-scale R4): the baked facts Parquet the server fold reads, relative to
   *  the version dir. The client never fetches it — it is the server's input. */
  factsParquet?: string;
  /** Group-regime only (companion-scale R4): the planner's fold-execution route, priced at bake. The
   *  client folds locally over companion planes when `client`, or posts to /api/views/{id}/fold when
   *  `remote` (behind the same seam). A `?fold=remote|client` query override wins over this. */
  foldRoute?: FoldRoute;
  /** Render-tile packaging (tile-transfer Phase 3): every tile's columns live as independently compressed
   *  blocks in one per-version archive, range-read per (tile, group), so a first paint ships geometry plus
   *  the active colour channel and never the measure planes it will not read. Absent selects the per-tile
   *  `z/x/y.arrow` path (formats 1/2, older bakes). Present ⇒ NO per-tile .arrow or .arrow.br exists. */
  renderPack?: RenderPack;
}

/** The render archive and its directory (mirror of Colossus.Domain.Model.RenderPack). `entries` maps a tile
 *  key (`z/x/y`) to each group's `[offset, length]`; a group is one column, except `@geom` which carries the
 *  encoded geometry payload (area marks) or the x/y pair (point marks). `firstPaint` is the group run a
 *  default paint needs, in pack order — the writer lays those blocks down adjacently at the head of each
 *  tile's span, so fetching exactly that set is ONE coalesced range request per tile. */
export interface RenderPack {
  file: string;
  codec: PackCodec;
  entries: Record<string, Record<string, [number, number]>>;
  firstPaint: string[];
  dict?: string;
  dictHash?: string;
}

/** The geometry group's reserved name — not a channel name (mirrors RenderPack.GeomGroup). */
export const GEOM_GROUP = '@geom';

/** The bake-priced fold-execution route (companion-scale R4). The client obeys `execution`; the rest are
 *  diagnostics (measured per-interaction bytes vs the budget). */
export interface FoldRoute {
  execution: 'client' | 'remote';
  worstTileBytes?: number;
  viewportEstimateBytes?: number;
  budgetBytes?: number;
  forced?: boolean;
}

/** One companion grain channel as a slab axis (SLAB-FORMAT §1–2). `domain` is the full ordered value list
 *  (categorical: canonical dict order; ordered: the sorted bin list). Axes are in cell order (fastest last). */
export interface SlabAxis {
  name: string;
  kind: 'categorical' | 'ordered';
  cardinality: number;
  cumulative: boolean;
  domain: string[];
}

/** Group-regime slab companion metadata (SLAB-FORMAT). `layout` is the view default; a tile's actual layout
 *  is `tileLayouts[key] ?? layout` (SLAB-FORMAT §3). `partials` names the planes and their element type. */
export interface CompanionSlab {
  layout: 'sparse' | 'dense';
  cells: number;
  occupancy: number;
  axes: SlabAxis[];
  partials: { name: string; type: 'f32' | 'i32' }[];
  /** Per-leaf-tile layout overrides: `tileKey → 'dense'|'sparse'` for tiles whose own occupancy disagrees
   *  with `layout`. Only exceptions are listed; absent tiles (and an absent field) use `layout`. */
  tileLayouts?: Record<string, 'sparse' | 'dense'>;
}

/** A tile's layout: its per-tile override if any, else the view default (SLAB-FORMAT §3). The one place the
 *  client resolves a per-tile choice — it must know a tile's layout before it fetches (sparse fetches the
 *  `@idx` structure, dense fetches `cnt`; the slice math differs). */
export function tileLayoutOf(slab: CompanionSlab, tileKey: string): 'sparse' | 'dense' {
  return slab.tileLayouts?.[tileKey] ?? slab.layout;
}

/** Companion pack block codec. "gzip" decodes with the browser-native `DecompressionStream`; "zstd" (a slab
 *  bake, Work Item C) with a WASM decoder + the pack's trained dictionary (compression lives inside the
 *  archive — a `Content-Encoding` wouldn't compose with range requests). */
export type PackCodec = 'gzip' | 'zstd';

/** The leaf companion archive and its directory: `file` is relative to the version directory, and `entries`
 *  maps a tile key `z/x/y` to its `[offset, length]` byte range. */
export interface CompanionPack {
  file: string;
  codec: PackCodec;
  entries: Record<string, [number, number]>;
  /** Trained shared dictionary path (relative to the version dir), for the "zstd" codec: fetched once per
   *  (view, version) and loaded into the client's zstd decoder. Absent for gzip, or a bake too small to train
   *  one (then zstd blocks decode dictionary-free). */
  dict?: string;
  /** SHA-256 (hex) of the dictionary file. */
  dictHash?: string;
  /** "row" (R2 leaf pack) or "slab" (R1). Absent ⇒ "row" (older bakes). */
  format?: 'row' | 'slab';
  /** Slab only: `tileKey → (planeName → [offset, length])`. Plane `"@idx"` is the CSR structure block;
   *  the rest are partial planes. A dense plane's region is its cell-row blocks concatenated. Enables
   *  plane-split fetch (R5); absent ⇒ whole-tile fetch. */
  planeEntries?: Record<string, Record<string, [number, number]>>;
  /** Cell-run slice directory (R5 second half): `tileKey → (planeName → per-cell-row block lengths)`, for
   *  **dense** tiles only (sparse opts out). A dense plane is one compressed block per cell row; cell `c`'s
   *  block is at `planeEntries[tile][plane][0] + Σ_{i<c} lengths[i]` for `lengths[c]` bytes, so only lengths
   *  are stored. Blocks are raw little-endian f32/i32 (per the partial), not Arrow. Absent (or a sparse tile)
   *  ⇒ whole-block fetch. */
  sliceEntries?: Record<string, Record<string, number[]>>;
}

const TILES_BASE = import.meta.env.VITE_TILES_BASE ?? 'http://localhost:5174/tiles';
export const API_BASE = import.meta.env.VITE_API_BASE ?? TILES_BASE.replace(/\/tiles\/?$/, '/api');

export function tileUrl(viewId: string, version: string, z: number, x: number, y: number): string {
  return `${TILES_BASE}/${viewId}/${version}/${z}/${x}/${y}.arrow`;
}

/** The fact companion beside a group-regime render tile. */
export function factsUrl(viewId: string, version: string, z: number, x: number, y: number): string {
  return `${TILES_BASE}/${viewId}/${version}/${z}/${x}/${y}.facts.arrow`;
}

/** The companion pack archive, tagged with the tile whose block a range request is after. Static
 *  servers ignore the query; it makes the URL — and so the service worker's cache key — per-tile. */
export function packBlockUrl(viewId: string, version: string, file: string, key: string): string {
  return `${TILES_BASE}/${viewId}/${version}/${file}?tile=${key}`;
}

/** The trained zstd dictionary URL for a (view, version) pack, or undefined when the bake trained none
 *  (gzip, or too few blocks). The client fetches it once and loads it into the worker's zstd decoder. */
export function packDictUrl(viewId: string, version: string, pack: { dict?: string } | undefined): string | undefined {
  return pack?.dict ? `${TILES_BASE}/${viewId}/${version}/${pack.dict}` : undefined;
}

/** Resolve latest.json → manifest.json for a view. */
export async function loadManifest(viewId: string): Promise<Manifest> {
  const pointer = await fetch(`${TILES_BASE}/${viewId}/latest.json`, { cache: 'no-cache' }).then((r) => r.json());
  const version: string = pointer.version;
  return fetch(`${TILES_BASE}/${viewId}/${version}/manifest.json`).then((r) => r.json());
}

File: src/Colossus.Domain/Model/Manifest.cs
using System.Text.Json.Serialization;

namespace Colossus.Domain.Model;

public readonly record struct TileMeta(int Z, int X, int Y, long Count, bool IsLeaf)
{
    [JsonIgnore] // derived — serializing it doubled every tile entry in manifest.json
    public TileId Id => new(Z, X, Y);
}

/// <summary>The boot document a client reads (via latest.json) to render a baked view.</summary>
public sealed record Manifest
{
    public int SchemaVersion { get; init; } = 1;
    /// <summary>Tile binary format. 2 = the zero-copy contract (single record batch, no nulls,
    /// tile-global triangle indices, canonical dictionaries, f32 measures) the client decodes as
    /// typed-array views over the one fetched buffer. Absent/0/1 on older manifests selects the
    /// copy-based format-1 decode, which the client keeps until every view is re-baked.</summary>
    public int TileFormat { get; init; }
    public required string Version { get; init; }
    public required ViewConfig View { get; init; }
    /// <summary>The reduction the planner chose for this bake (derived from data shape, not authored).
    /// The client reads it to label fidelity — e.g. an aggregate pyramid marks its coarse levels.</summary>
    public required ReductionKind Reduction { get; init; }
    public required string Regime { get; init; }
    public required Bbox Root { get; init; }
    public required int MinZoom { get; init; }
    public required int MaxZoom { get; init; }
    public required int TilePointBudget { get; init; }
    public required long TotalPoints { get; init; }
    /// <summary>Rows the source reported at bake time. <see cref="TotalPoints"/> is *defined* as the leaf
    /// sum, so only this — a number the reducer never produced — can witness that the tiling partitioned
    /// the source rather than duplicating or dropping rows. Null on manifests from older bakes.</summary>
    public long? SourceRows { get; init; }
    public required IReadOnlyList<TileMeta> Tiles { get; init; }
    /// <summary>Per-channel data domains scanned from the full staged extract at bake time, keyed by
    /// channel name. Spares the client a root-tile scan at load and — unlike that tile, which is a
    /// sample — sees every row, so no category can be missing. Null on manifests from older bakes;
    /// the client falls back to its tile scan.</summary>
    public IReadOnlyDictionary<string, ChannelDomain>? ChannelDomains { get; init; }

    /// <summary>Group regime only: the derived perMark/perFact split so the client can
    /// route a filter as a GPU predicate (perMark) or a fold context (perFact). Null in the row regime.</summary>
    public FactChannels? FactChannels { get; init; }

    /// <summary>True when the bake wrote a <c>z/x/y.facts.arrow</c> companion beside every tile.</summary>
    public bool CompanionTiles { get; init; }

    /// <summary>The companion grain columns (perFact dict + temporal channels), so the client knows what
    /// dimensions its fact partials are keyed by. Null in the row regime.</summary>
    public IReadOnlyList<string>? GrainChannels { get; init; }

    /// <summary>Companion packaging (companion-scale R2, extended by R1/R5). Every companion tile's blocks
    /// live in one per-version archive, fetched by HTTP range. Null selects the per-file layout (older
    /// bakes). A slab bake sets <see cref="CompanionPack.Format"/> = "slab" and packs both leaf and internal
    /// levels (internal compression, R5); a row-form R2 bake leaves it "row" with leaf-only entries.</summary>
    public CompanionPack? CompanionPack { get; init; }

    /// <summary>Group-regime slab companion metadata (companion-scale R1). Null on row-form bakes — the
    /// client keeps the row-form decode/fold. Present ⇒ the pack's blocks are slab planes.</summary>
    public CompanionSlab? CompanionSlab { get; init; }

    /// <summary>Render-tile packaging (tile-transfer Phase 3). Every tile's columns live as independently
    /// compressed blocks in one per-version archive, fetched by HTTP range, so a first paint ships geometry
    /// and the active colour channel alone and never the measure planes it will not read. Null selects the
    /// per-tile file layout (formats 1/2 and older bakes) — the same gating pattern
    /// <see cref="CompanionPack"/> uses. Present ⇒ no per-tile <c>.arrow</c>/<c>.arrow.br</c> exists.</summary>
    public RenderPack? RenderPack { get; init; }

    /// <summary>Group-regime only (companion-scale R4): the baked facts Parquet, relative to the version
    /// directory, that the server fold reads (DuckDB over the baked artifact, never the source DB — RULES
    /// R5). Retained per version so a fold is reproducible against exactly the bake it prices. Null in the
    /// row regime and on older group bakes that predate R4.</summary>
    public string? FactsParquet { get; init; }

    /// <summary>Group-regime only (companion-scale R4): the planner's fold-execution route, priced at bake
    /// from measured companion bytes. The client reads <see cref="FoldRoute.Execution"/> to pick the local
    /// fold or the remote fold endpoint behind the same seam. Null in the row regime.</summary>
    public FoldRoute? FoldRoute { get; init; }
}

/// <summary>The planner's fold-execution decision for a group-regime view (companion-scale R4). Priced at
/// bake from the measured per-interaction companion transfer (plane-split bytes, R5): a view whose worst
/// leaf tile exceeds the budget routes its fold to the server, shipping folded columns instead of facts.
/// The extra fields are diagnostics — the client obeys only <see cref="Execution"/>.</summary>
public sealed record FoldRoute
{
    /// <summary>"client" (fold in the browser over fetched companion planes) or "remote" (fold on the
    /// server over the baked facts Parquet, ship folded columns).</summary>
    public required string Execution { get; init; }
    /// <summary>Measured worst single-tile per-interaction transfer (the color measure's plane-split bytes)
    /// — the dense-leaf number the budget guards.</summary>
    public long WorstTileBytes { get; init; }
    /// <summary>Measured estimate for a dense screenful (sum of the largest N tiles' per-interaction bytes).</summary>
    public long ViewportEstimateBytes { get; init; }
    /// <summary>The configured budget the worst tile was compared against.</summary>
    public long BudgetBytes { get; init; }
    /// <summary>True when the route was forced remote by config regardless of price (benchmark/testing).</summary>
    public bool Forced { get; init; }
}

/// <summary>The companion archive and its directory. <see cref="File"/> is relative to the version
/// directory; <see cref="Codec"/> is an encoding the browser-native <c>DecompressionStream</c> accepts;
/// <see cref="Entries"/> maps a tile key (<c>z/x/y</c>) to the <c>[offset, length]</c> byte range of its
/// whole region — compression lives inside the archive because <c>Content-Encoding</c> doesn't compose
/// with ranges. A slab bake (<see cref="Format"/> = "slab") additionally fills <see cref="PlaneEntries"/>
/// with per-plane ranges so a fold fetches only the planes its active measures need (R5 plane split).</summary>
public sealed record CompanionPack
{
    public required string File { get; init; }
    /// <summary>Block codec: "gzip" (browser-native <c>DecompressionStream</c>) or "zstd" (companion-scale
    /// Work Item C — high-level zstd with a trained shared dictionary; the client uses a WASM zstd decoder).</summary>
    public required string Codec { get; init; }
    public required IReadOnlyDictionary<string, long[]> Entries { get; init; }

    /// <summary>Trained shared dictionary path (relative to the version directory), for the "zstd" codec: one
    /// dictionary per (view, version) over sampled blocks, so the small sliced cell-row blocks compress well
    /// (companion-scale R5). The client fetches it once and loads it into its zstd decoder. Null when the
    /// codec needs no dictionary (gzip, or a bake with too few blocks to train one).</summary>
    public string? Dict { get; init; }

    /// <summary>SHA-256 (hex) of the dictionary file — recorded so the client can key/verify it. Null with
    /// <see cref="Dict"/>.</summary>
    public string? DictHash { get; init; }

    /// <summary>"row" (R2 leaf pack of row-form companions) or "slab" (R1). Absent ⇒ "row" (older bakes).</summary>
    public string? Format { get; init; }

    /// <summary>Slab only: <c>tileKey → (planeName → [offset, length])</c>. Plane <c>"@idx"</c> is the CSR
    /// structure block (sparse layout); the rest are partial planes. A dense plane's region is the
    /// concatenation of its per-cell-row blocks (<see cref="SliceEntries"/>); a whole-plane fetch ranges the
    /// region and inflates each block. Null ⇒ whole-tile fetch only.</summary>
    public IReadOnlyDictionary<string, IReadOnlyDictionary<string, long[]>>? PlaneEntries { get; init; }

    /// <summary>Cell-run slice directory (companion-scale R5 second half): <c>tileKey → (planeName → per-cell
    /// compressed block lengths)</c>, present only for **dense** tiles (sparse opts out — SLAB-FORMAT §5). A
    /// dense plane is stored as one independently compressed block per cell row (all marks at one cell — the
    /// slice unit, §4b); cell <c>c</c>'s block is at <c>PlaneEntries[tile][plane][0] + Σ_{i&lt;c} lengths[i]</c>
    /// for <c>lengths[c]</c> bytes, so only the block lengths are recorded. The client fetches only the cell
    /// rows the active context reads; absence of this directory (or a sparse tile) selects the whole-block
    /// fetch. Blocks are raw little-endian typed-array bytes (f32/i32 per the partial), not Arrow — per-row
    /// Arrow framing would swamp a small tile's payload.</summary>
    public IReadOnlyDictionary<string, IReadOnlyDictionary<string, int[]>>? SliceEntries { get; init; }
}

/// <summary>The render-tile archive and its directory (tile-transfer Phase 3). One archive per (view,
/// version) replacing the per-tile render files; <see cref="File"/> is relative to the version directory.
/// Compression lives inside the archive, per block, because <c>Content-Encoding</c> does not compose with
/// Range requests.
///
/// <para><b>Block order is the design.</b> A tile's blocks are laid down in first-paint order — the
/// geometry group, then the default colour channel, then the filter-slot channels, then everything else —
/// so the default first paint is ONE contiguous byte range per tile and a whole-tile read is one range over
/// the tile's full span. The order is derived from the view's declared channel roles at bake time, never
/// authored and never keyed on channel names. Future formats must preserve this property the way the slab
/// format's cell order is preserved.</para></summary>
public sealed record RenderPack
{
    public required string File { get; init; }

    /// <summary>Block codec: "zstd" (with the trained shared <see cref="Dict"/>) or "gzip" (browser-native
    /// <c>DecompressionStream</c>).</summary>
    public required string Codec { get; init; }

    /// <summary><c>tileKey → (group → [offset, length])</c>. A group is one column, except the geometry
    /// group (<see cref="GeomGroup"/>) which carries the encoded geom3 payload for area marks or the
    /// representative x/y columns for point marks. Each block is a standalone single-batch Arrow IPC
    /// stream, so a column decodes alone and stays a zero-copy view over its own inflated buffer.</summary>
    public required IReadOnlyDictionary<string, IReadOnlyDictionary<string, long[]>> Entries { get; init; }

    /// <summary>The groups a first paint needs, in pack order: geometry, the default colour channel, and the
    /// filter-slot channels. Recorded rather than re-derived so the client's request set is exactly the run
    /// the writer laid out contiguously. Every other group is fetched lazily — on a colour switch or a
    /// click.</summary>
    public required IReadOnlyList<string> FirstPaint { get; init; }

    /// <summary>Trained shared dictionary path (relative to the version directory), for the "zstd" codec —
    /// one per (view, version) over sampled blocks, so small per-column blocks still compress well. Null
    /// when the codec needs none (gzip, or too few blocks to train one).</summary>
    public string? Dict { get; init; }

    /// <summary>SHA-256 (hex) of the dictionary file, so the client can key/verify it. Null with
    /// <see cref="Dict"/>.</summary>
    public string? DictHash { get; init; }

    /// <summary>The geometry group's reserved name. Not a channel name — the one group whose contents are
    /// decided by the mark, not by the view's channel list.</summary>
    public const string GeomGroup = "@geom";

    /// <summary>The channels a first paint needs, in block order, derived purely from the view's declared
    /// roles: the colour channel (what the default paint reads) followed by the filter-slot channels (the
    /// GPU filter attribute is built per tile at decode, so its columns must be resident too). Everything
    /// else — the other measures, the identity/inspect channels — is lazy. No channel name and no dataset
    /// shape appears here; a view that declares neither role simply has a geometry-only first paint.</summary>
    public static IReadOnlyList<string> FirstPaintChannels(ViewConfig view)
    {
        var ordered = new List<string>();
        void Add(string? name)
        {
            if (!string.IsNullOrEmpty(name) && !ordered.Contains(name, StringComparer.Ordinal)) ordered.Add(name);
        }
        Add(view.Encoding?.Color?.Channel);
        foreach (var f in view.Filters ?? []) Add(f.Channel);
        return ordered;
    }
}

/// <summary>The derived channel split of a group-regime view.</summary>
public sealed record FactChannels(IReadOnlyList<string> PerMark, IReadOnlyList<string> PerFact);

/// <summary>One channel's observed domain. Numeric channels carry min/max plus a quantile grid (the
/// client derives quantile-scale breaks from it); non-numeric channels carry their distinct values,
/// capped — a capped list sets <see cref="ValuesTruncated"/> and the client treats it as absent.</summary>
public sealed record ChannelDomain
{
    public IReadOnlyList<string>? Values { get; init; }
    public bool? ValuesTruncated { get; init; }
    public double? Min { get; init; }
    public double? Max { get; init; }
    public IReadOnlyList<double>? Quantiles { get; init; }
}

/// <summary>The single mutable pointer, flipped atomically once a bake commits.</summary>
public sealed record LatestPointer(string Version);
